### PR TITLE
Feat add - Added  4 builders, their extension methods, and the appropriate testing

### PR DESCRIPTION
**Pull request by LukeHarjulin**
### Builders and their extensions:
**Network Security Group**
Extensions:
- Name
- Resource Group
- Location

**Security Rule**
- Name
- Resource Group
- Network Security Group
- Direction
- Source port range
- Destination port range
- Source port ranges
- Destination port ranges
- Priority
- Access
- Source prefix
- Destination prefix
- Source prefixes
- Destination prefixes
- Source Application Security Group
- Destination Application Security Group
- Description

**Network Interface**
- Name
- Resource Group
- Location
- Network Security Group
- IP Configuration Name
- IP Configuration Subnet ID
- IP Configuration Address Version
- IP Configuration Allocation Method

**Application Security Group**
- Name
- Resource Group
- Location

### CODE REVIEW

**Comment by sesispla** [comment]
Thank you @LukeHarjulin please note this branch has conflicts. 

Could you please merge "main" against your branch, solve all conflicts and push the changes back again?

**Comment by LukeHarjulin** [comment]
@sesispla Sorry about the conflicts. I have resolved all conflicts and passed all checks for the feat-add branch, merge should be possible now.

File: src/Infrastructure/Stize.Infrastructure.Azure/Networking/ApplicationSecurityGroupExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        /// appends the hex value of the RandomId to the end of the name
+        /// </summary>
+        /// <param name="builder">Builder instance</param>
+        /// <param name="name">Builder name</param>
+        /// <returns>The builder argument</returns>
+        public static ApplicationSecurityGroupBuilder Name(this ApplicationSecurityGroupBuilder builder, Input<string> name)
+        {
+            if (builder.RandomId != null)
+            {
+                builder.Arguments.ApplicationSecurityGroupName = builder.RandomId.Hex.Apply(r => $"{name}-{r}");
```
Comment (sesispla) [inline src/Infrastructure/Stize.Infrastructure.Azure/Networking/ApplicationSecurityGroupExtensions.cs:31]: I recently found that the logic for creating the random names was wrong. Please find a corrected example:

https://github.com/stize/infrastructure/blob/0122c548778a1a6789fd91ca8b3368be2f6af32b/src/Infrastructure/Stize.Infrastructure.Azure/ResourceGroupExtensions.cs#L18

Use the same approach here please.

The reason being, name and Hex are both Pulumi Input values. We need to Apply both (in a concatenation) in order for the evaluated values to be passed to the resource name.

File: src/Infrastructure/Stize.Infrastructure.Azure/Networking/NetworkSecurityGroupExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        /// appends the hex value of the RandomId to the end of the name
+        /// </summary>
+        /// <param name="builder">Builder instance</param>
+        /// <param name="name">Builder name</param>
+        /// <returns>The builder argument</returns>
+        public static NetworkSecurityGroupBuilder Name(this NetworkSecurityGroupBuilder builder, Input<string> name)
+        {
+            if (builder.RandomId != null)
+            {
+                builder.Arguments.NetworkSecurityGroupName = builder.RandomId.Hex.Apply(r => $"{name}-{r}");
```
Comment (sesispla) [inline src/Infrastructure/Stize.Infrastructure.Azure/Networking/NetworkSecurityGroupExtensions.cs:33]: Same here. I recently found that the logic for creating the random names was wrong. Please find a corrected example:

https://github.com/stize/infrastructure/blob/0122c548778a1a6789fd91ca8b3368be2f6af32b/src/Infrastructure/Stize.Infrastructure.Azure/ResourceGroupExtensions.cs#L18

Use the same approach here please.

The reason being, name and Hex are both Pulumi Input values. We need to Apply both (in a concatenation) in order for the evaluated values to be passed to the resource name.

File: src/Infrastructure/Stize.Infrastructure.Azure/Networking/SecurityRuleExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        /// appends the hex value of the RandomId to the end of the name
+        /// </summary>
+        /// <param name="builder"><see cref="SecurityRule"/> builder</param>
+        /// <param name="name"><see cref="SecurityRule"/> name</param>
+        /// <returns></returns>
+        public static SecurityRuleBuilder Name(this SecurityRuleBuilder builder, Input<string> name)
+        {
+            if (builder.RandomId != null)
+            {
+                builder.Arguments.SecurityRuleName = builder.RandomId.Hex.Apply(r => $"{name}-{r}");
```
Comment (sesispla) [inline src/Infrastructure/Stize.Infrastructure.Azure/Networking/SecurityRuleExtensions.cs:21]: And here. I recently found that the logic for creating the random names was wrong. Please find a corrected example:

https://github.com/stize/infrastructure/blob/0122c548778a1a6789fd91ca8b3368be2f6af32b/src/Infrastructure/Stize.Infrastructure.Azure/ResourceGroupExtensions.cs#L18

Use the same approach here please.

The reason being, name and Hex are both Pulumi Input values. We need to Apply both (in a concatenation) in order for the evaluated values to be passed to the resource name.

**Review by sesispla — CHANGES_REQUESTED** [review]
Please address all the unit testing before completing this PR.

See how Unit Testing has been solved in the main branch:

How to run a test:
https://github.com/stize/infrastructure/blob/0122c548778a1a6789fd91ca8b3368be2f6af32b/src/Infrastructure/Stize.Infrastructure.Azure.Tests/ResourceGroupTests.cs#L37-L59

How the stack should look like:
https://github.com/stize/infrastructure/blob/0122c548778a1a6789fd91ca8b3368be2f6af32b/src/Infrastructure/Stize.Infrastructure.Azure.Tests/Stacks/RandomIdNameStack.cs#L9-L27

And finally, how the Pulumi Engine mock can potentially look like (open to suggestions, improvements!)

https://github.com/stize/infrastructure/blob/0122c548778a1a6789fd91ca8b3368be2f6af32b/src/Infrastructure/Stize.Infrastructure.Azure.Tests/Stacks/RandomIdNameMock.cs#L7-L50

The most challenging part is, without a doubt, the Pulumi Engine Mock.

**Review by sesispla — APPROVED** [review]
LGTM

**Pull request merged**